tenders/spt/spt_core.c

```diff
@@ -174,6 +174,12 @@ void spt_run(struct spt *spt, uint64_t p_entry)
     uint64_t sp = spt->mem_size - 0x8;
 #elif defined(__aarch64__)
     uint64_t sp = spt->mem_size - 0x10;
+#elif defined(__powerpc64__)
+    /*
+     * stack alighment is 0x10; we need an additional 0x8 so _start can
+     * store the link register to 16(r1) (r1 = stack pointer)
+     */
+    uint64_t sp = spt->mem_size - 0x20;
```

Comment:
Continuing the conversation from https://github.com/Solo5/solo5/pull/368#discussion_r291230236 regarding the setup of the stack frame and calling convention here. I wanted to understand exactly what PPC64 is doing, so went searching and read through the related sections of the following documents:

1. 64-bit PowerPC ELF Application Binary Interface Supplement 1.9, Section 3.2.2, https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi-1.9.html#STACK
2. Function calls and the PowerPC 64-bit ABI, Table 1. Stack frame layout, https://www.ibm.com/developerworks/library/l-powasm4/index.html

It turns out that the minimum stack frame size on PPC64 is 112 bytes (minimum 64 byte parameter save area, plus 48 bytes for the other save areas below that) and this (unlike other architectures I have experience with) **must be allocated by the caller**, while maintaining stack pointer alignment to a quadword (0x10).

So, I think the above should read:

```
    /*
     * Stack alignment on PPC64 is 0x10, minimum stack frame size is 112 bytes.
     */
    uint64_t sp = spt->mem_size - 112;
```

Additionally, you might want to store 0x0 at (r1) in spt_launch itself before calling _start, in case something changes up the chain and we pass in uninitialised memory. This is so that the back chain for the first stack frame is always 0 (NULL).

Does that clear things up?

Reply:
2) seems to be the better documentation. I'll adjust it.